src/v/cloud_storage/partition_manifest.cc

```diff
@@ -263,6 +265,61 @@ std::optional<model::offset> partition_manifest::get_start_offset() const {
     return _start_offset;
 }
 
+std::optional<kafka::offset>
+partition_manifest::get_start_kafka_offset() const {
+    if (_start_offset != model::offset{}) {
+        auto iter = _segments.find(_start_offset);
+        if (iter != _segments.end()) {
+            auto delta = iter->second.delta_offset;
+            return _start_offset - delta;
+        } else {
+            throw std::runtime_error(fmt_with_ctx(
+              fmt::format,
+              "can't translate start offset {} because the manifest is empty",
+              _start_offset));
+        }
+    }
+    return std::nullopt;
+}
+
+partition_manifest::const_iterator
+partition_manifest::segment_containing(kafka::offset o) const {
+    vlog(cst_log.debug, "Metadata lookup using kafka offset {}", o);
+    // Kafka offset is always <= log offset.
+    // To find a segment by its kafka offset we can simply query
+    // manifest by log offset and then traverse forward until we
+    // will find matching segment.
+    auto it = segment_containing(kafka::offset_cast(o));
+    auto prev = end();
+    while (it != end()) {
+        auto base = it->second.base_offset - it->second.delta_offset;
+        if (base > o) {
+            // We need to find first element which has greater kafka
+            // offset then the target and step back. It is possible
+            // to have a segment that doesn't have data batches. This
+            // scan has to skip segments like that.
+            break;
+        }
+        prev = it;
+        it = std::next(it);
+    }
+    if (it == end() && prev != end()) {
+        if (prev->second.delta_offset_end != model::offset_delta{}) {
+            // In case if 'prev' points to the last segment it's not guaranteed
+            // that the segment contains the required kafka offset. We need an
+            // extra check using delta_offset_end. If the field is not set then
+            // we will return the last segment. This is OK since
+            // delta_offset_end will always be set for new segments.
+            auto m = prev->second.committed_offset
+                     - prev->second.delta_offset_end;
```

Comment:
> Does delta_offset_end always correspond to the delta at the committed_offset?

yes, it does

> I'm curious, what's the significance behind this commited_offset vs some end_offset or other? For a given remote segment, wouldn't they be the same, since we only upload committed data? 

We only uploading committed data and this name is used because when we're uploading a full segment we're using its `offsets().committed_offset` field. So basically it's both committed offset and end or last offset of the segment.

> And the reason we don't is because it is a newly introduced field?

correct